Fix segmentation fault when processing assert statements

mohsinm-dev opened the pull request:
Resolves issue #1442 where Clad would crash with a segfault when differentiating functions containing assert statements. The crash occurred in ReferencesUpdater::updateType() when processing SourceLocExpr nodes that return null QualType values.

Changes:
- Add null checks in ReferencesUpdater::VisitStmt() before calling updateType()
- Add defensive null check in ReferencesUpdater::VisitDeclRefExpr() for consistency
- Add test case to prevent regression

Fixes #1442

mohsinm-dev committed:
Fix segmentation fault when processing assert statements

Resolves issue #1442 where Clad would crash with a segfault when
differentiating functions containing assert statements. The crash
occurred in ReferencesUpdater::updateType() when processing
SourceLocExpr nodes that return null QualType values.

Changes:
- Add null checks in ReferencesUpdater::VisitStmt() before calling updateType()
- Add defensive null check in ReferencesUpdater::VisitDeclRefExpr() for consistency
- Add test case to prevent regression

Fixes #1442

vgvassilev commented on lib/Differentiator/StmtClone.cpp:
```diff
@@ -564,13 +564,18 @@ bool ReferencesUpdater::VisitDeclRefExpr(DeclRefExpr* DRE) {
+  if (auto* E = dyn_cast<Expr>(S)) {
+    QualType QT = E->getType();
+    if (!QT.isNull())
+      updateType(QT);
```
Should we move the `isNull` check in `updateType`?

mohsinm-dev replied:
Yes, I think moving `isNull` check in `updateType` is better. It will be a single point of validation. What do you think ?

mohsinm-dev replied:
Let me know if you’ve moved it or if you want me to update it. I’m happy to do that.

vgvassilev replied:
I have not done anything. Please move it.

vgvassilev commented on test/Misc/AssertFix.C:
```diff
@@ -0,0 +1,24 @@
+// RUN: %cladclang %s -I%S/../../include -fsyntax-only -Xclang -verify 2>&1 | %filecheck %s
```
I'd move that test into the `Regression` folder under some `issue-xxx.cpp`...

mohsinm-dev commented:
> @mohsinm-dev, it seems we are almost ready to go. We need to capture the warning that comes with the test cases as an expected warning...

Yes, are these warnings the reason for failed workflows? Let me know what we have to do. I'll look into it.

vgvassilev commented:
> > @mohsinm-dev, it seems we are almost ready to go. We need to capture the warning that comes with the test cases as an expected warning...
> 
> Yes, are these warnings the reason for failed workflows? Let me know what we have to do. I'll look into it.

I've adjusted the test for osx, however, it looks like the current fix is not enough for ubuntu. Can you take a look?

vgvassilev commented:
Your last push undid the changes I made to your branch to fix osx.

vgvassilev commented:
> > Your last push undid the changes I made to your branch to fix osx.
> 
> Yes, I was trying something. I am working on it. Will push the updates soon. Hopefully that will resolve it.

If you need to debug the runners here is how: https://clad.readthedocs.io/en/latest/user/DevelopersDocumentation.html#debugging-github-runners

mohsinm-dev commented:
> > > Your last push undid the changes I made to your branch to fix osx.
> > 
> > 
> > Yes, I was trying something. I am working on it. Will push the updates soon. Hopefully, that will resolve it.
> 
> If you need to debug the runners here is how: https://clad.readthedocs.io/en/latest/user/DevelopersDocumentation.html#debugging-github-runners

Can you guide me a bit? I tried running the debugger, but was unable to connect. I guess there might be some internal fixes that need to be done. I am looking into that and debugging it. Probably will try to wrap up this weekend.

vgvassilev commented:
Ok, we are getting there. Older versions fail because it seems we might not be handling properly the PredefinedExpr expression kind. Perhaps you can build locally an older version of clang and debug?

mohsinm-dev commented:
> Ok, we are getting there. Newer versions fail because it seems we might not be handling properly the PredefinedExpr expression kind. Perhaps you can build locally a newer version of clang and debug?

You're right about the PredefinedExpr issue. I found that __func__, __FUNCTION__, and __PRETTY_FUNCTION__ return null QualType values, causing segfaults in CloneType(). Fixed by adding null check, also updated the test to cover PredefinedExpr cases.

mohsinm-dev committed:
Fix PredefinedExpr segfault in CloneType

Add null check to prevent crashes when processing __func__, __FUNCTION__,
and __PRETTY_FUNCTION__ expressions. Tested with clang 20.1.8.

mohsinm-dev committed:
Make test platform-independent by removing cassert dependency

- Remove #include <cassert> to avoid platform-specific implementations
- Define custom TEST_ASSERT macros that simulate different assert variants
- Test both __FILE__/__func__ and __builtin_FILE()/__builtin_FUNCTION() versions
- Ensures consistent behavior across Ubuntu and macOS workflows

vgvassilev commented on test/Regressions/issue-1442.cpp:
```diff
@@ -0,0 +1,39 @@
+// This addresses issue #1442
+
+#include "clad/Differentiator/Differentiator.h"
+#include <cassert>
```
We don't need a custom header, we just need to pull in the relevant parts of the `cassert` header to make the example standalone. Do you think you can do that?

mohsinm-dev committed:
Add SourceLocExpr cloning and fix test for cross-platform CI

- Add SourceLocExpr to StmtClone so __builtin_FILE/__builtin_LINE/
  __builtin_FUNCTION (used by newer glibc assert) can be cloned properly
- Replace __assert_fail with custom test_assert_handler in test to avoid
  conflicting with glibc's declaration brought in via Differentiator.h's
  #include <cassert> (glibc uses unsigned int for line, test had int)
- Use unsigned int for line parameter to match __builtin_LINE() return type
